---
title: "COMP90014 - Assignment2"
author: "Michael McLellan - 665968"
date: "Due Date; 17 October 2014"
output:
  word_document: default
  pdf_document:
    fig_caption: yes
---

```{r global-options, include=F}
knitr::opts_chunk$set(fig.height=4, fig.width=5, tidy=TRUE, warning=FALSE, message=FALSE)
load(file="Assignment2.RData")
library(edgeR)
library(knitr)
```

## Task 1 - Differential gene expression analysis 
For the following analyses the R package edgeR was used to find differential expressed genes between a male and female group with each group containing twenty replicants. Initial set up for the experiment is to load the data into R Studio and look at the data. The data is an RNA-Seq data set containing only the counts with the row names as the gene ids and the column names as sample ids. 
```{r, eval=FALSE}
#Load in read in counts table.
alldata = read.table("assignment2data_fullcounts.tsv", header=TRUE, sep="\t")
```

A good place to start for an analyses is too look at the data. In this analysis we look at the dimension of the data, showing how many genes and individuals we are working with. The output shows that there are 40 individuals and 44061 genes in the data set as we expected. The data is also in the correct format for edgeR.
```{r, results='asis'}
#Look at dimension, row and column names
dim(alldata)
```

First step is to filter out genes with zeroes in both the male and female repliants. For this analysis it was decided to remove genes with zero counts if both the female and male samples have predominantly zeros. Using the apply function we filter out genes who's medians in both the male and female samples are zero. As a result 19443 genes were removed. 
```{r, eval=FALSE}
#Apply function to remove genes whos median is 0 for both male 
#and female samples and view new dimension of filtered data set.
low_counts = apply(alldata[,1:20],1,median)==0 & apply(alldata[,21:40],1,median)==0
filtered = alldata[!low_counts,]
```
```{r}
dim(filtered)
```

Next the data was split into two data frames. One containing the raw counts and the other containing the information for the gene information.
```{r,eval=FALSE}
#Split data into expression counts and gene information
expression = filtered[,1:40]
geneInfo = filtered[,41:44]
```

Now we start to use edgeR to perform a differential expression analysis on the data set. 

We are required to provide edgeR with the information telling which of the data belongs to each group. The standard way to do this is to use a design matrix, but as this is a simple two case comparison we only need to provide edgeR with a vector. The 'ismale' vector, 0's represent the columns belonging to the female group while 1's being to the male group. Next we package all the counts into a list object called DGEList for edgeR. The DGEList function converts the count matrix into an edgeR object. The counts and information can still be accessed with in the DGE object. 
```{r, eval=FALSE}
#Vector distinguishing which group the twe belong too. 
ismale = c(rep(0,20), rep(1,20))
ismale
#Creating the DGE object
dge = DGEList(expression, group=ismale)
```

To take into account variation as a result of the different experiments from the replicates,  edgeR's built in normalisation is used to scale each samples counts. By default a TMM (trimmed mean) normalisation is used. 
```{r,eval=FALSE}
#Normalising the data.
dge = calcNormFactors(dge)
```

A good step in an analysis is to visualise the data. We can use a multi-dimensional scaling (MDS) plot to visualise the level of similarity of the two cases in the data. The basic version of a MDS plot is known as Principle coordinates analysis. It can clearly be seen that there is clustering between the male and female groups. 
```{r, fig.cap="The multidimensional scaling plot is a way to visualise the similarity between the individuals in the data set. We can clearly see the male and females clustering into seperate groups. ", message=FALSE,}
plotMDS(dge, col=c( rep("red",20), rep("blue",20)))
```

Next we model the variance of the data. Using edgeR we fit the negative binomial model to find the common dispersion for the data, so each gene has the same value for the dispersion. Next the per-gene estimates of variance are found using the "estimateTagewiseDisp" function. The estimates are computed using an empirical Bayes method on weighted conditional maximum likelihood. The model uses the common dispersion as a prior when calculating the negative binomial dispersion preventing over fitting to the data. The BCV plot (biological coefficient of variance) plots the biological coefficient of variation (BCV) against gene abundance showing the common the and tag wise BCV estimates. 
```{r, fig.cap="The BCV plot shows the common dispersion as the red line and the tagwise estimates as the black dots."}
#Estimate the common and tag wise dispersion, and view the relationship in a BCV plot
dge = estimateCommonDisp(dge)
dge$common.dispersion
dge = estimateTagwiseDisp(dge)
plotBCV(dge)
```

By plotting the mean-variance relationship we are able to see how well the data fits to the estimate dispersion parameters. The grey dots are the raw variance of the counts, blue dots are the tag wise dispersion and the solid blue line is the variance using the common dispersion. The straight line is the mean variance relationship of a Poisson distribution. The plot models the mean variance relationship of the estimated dispersion. The relationship is not linear but is probably quadratic.
```{r, fig.cap='The grey dots are the raw variance of the counts, blue dots are the tag wise dispersions and the solid blue line is the variance using the common dispersion.'}
plotMeanVar(dge, show.raw.vars=TRUE, show.tagwise.vars=TRUE, 
            show.binned.common.disp.vars=FALSE, show.ave.raw.vars=FALSE, 
            dispersion.method="qcml", NBline=TRUE, nbins=100, pch=16 , 
            xlab="Mean Expression (Log10 Scale)", ylab = "Variance (Log10 Scale)")
```

To test for DGE we perform pair-wise tests between the male and female groups. The exact test looks for statistically significant differences between the means of the negative binomial distribution fitted to the data.The results are stored in an edgeR data class 'DGEExact' with each row containing the gene with its logFC value, logCPM and P Value from the test.
```{r,eval=FALSE}
#Perform an exact test on the data and view the structure of the results
results = exactTest(dge)
str(results)
```

We can view the top ten tags from the results, IE the top ten differential expressed genes between the two groups. We can then refer back to the data frame containing the information for each gene and find the information for each differential expressed gene. 
```{r}
#Top ten differentially expressed genes show with gene information
topGI = merge(geneInfo[rownames(topTen),], topTags(results, n=10), by=0, sort=FALSE)
del = c("Row.names", "start_position", "FDR") #Columns not needed in output
topGI[,!(colnames(topGI) %in% del), drop=FALSE]
```

To view the bottom ten results we can get the full list from the `topTags()` command and then use the tail function in R to select the bottom ten rows of the data frame. For this the bottom ten results were interpreted to be the the bottom of the list of genes after being sorted. 
```{r}
#Bottom 10 genes from the data set after being sorted by the topTags function. 
low = topTags(results, n=nrow(results$table))$table
low_GI = merge(geneInfo[rownames(low),], tail(low, n=10), by=0, sort=FALSE)
low_GI[,!(colnames(low_GI) %in% del), drop=FALSE]
```

The decideTestsDGE() function will implement multiple testing procedures to determine whether each log-fold change in the results matrix should be considered significantly different from zero. The summary output shows the number of genes whose fold change is up or down. 
```{r}
de = decideTestsDGE(results)
summary(de)
```

For visualisation a heatmap is a common way to view expression data. A heatmap is a two-dimensional grid with different colours representing expression values. For this case we will use the top twenty differential expressed genes in the heatmap. The heatmap shows clear areas of different levels of expression. 
```{r, fig.cap='Heat map for the top 20 differentially expressed genes'}
#Produce heat map
topgenes = rownames(topTags(results, n=20))
log.dge = cpm(dge, prior.count=2, log=TRUE)
topgene.log.expression = log.dge[topgenes,]
redtogreen = colorRampPalette(c("red", "yellow", "green"))(n=299)
heatmap(topgene.log.expression, col=redtogreen)
```

A second common plot is a histogram plot showing the distribution of P-values. 
```{r, fig.cap='Histogram of Pvalues from the results table'}
hist(results$table$PValue, breaks=150)
```

Another useful plot is a smear plot that shows the relationship between concentration and fold change across the genes. The DE genes are coloured red while non-DE genes are coloured black. The orange dots represent genes with zero counts. The blue lines represent two times fold change in expression which can be viewed as biologically significant. 
```{r, fig.cap='Smear plot showing the fold changes for genes.'}
#Smear plot
resultsTbl = topTags(results, n=nrow(results$table))$table
de.genes = rownames(resultsTbl)[resultsTbl$PValue <= 0.05]
plotSmear(results, de.tags=de.genes)
abline(h=c(-2, 2), col="blue") 
```

## Task 1 Numerical Questions
1. Number of genes with p-values below 0.05, What proportion of the genes tested?
```{r, strip.white=FALSE}
nrow(results$table[results$table$PValue < 0.05, ])
(nrow(results$table[results$table$PValue < 0.05, ]) / nrow(results$table))
```

2. Out of all genes on Y chromosome, how many have a P-value below 0.05?
```{r, strip.white=FALSE}
significant = results$table[results$table$PValue < 0.05, ]
significantGI = geneInfo[rownames(significant),]
significantY = geneInfo[geneInfo$chromosome_name=="Y", ]
nrow(significantY)
```
3. Out of the 100 genes with the lowest P-value, how many are from the X chromosome?
```{r, strip.white=FALSE}
top100 = topTags(results, n=100)
significantX = geneInfo[rownames(top100), ]
nrow(significantX[significantX$chromosome_name=="X", ])
```
4. What is the log-fold change for the gene XIST?
```{r, strip.white=FALSE}
xist = geneInfo[geneInfo$gene_name=="XIST", ]
xistResult = results$table[rownames(results$table)==rownames(xist), ]
xistResult$logFC
```

### Task 2 - DGE analysis of subsampled dataset 
For the second task a differential expression analysis on a sub sample of the data set. The sub sample contained the same number of replicants and genes but had different number of counts. The same workflow was followed as task one.

The sub sample data set was loaded, genes with zero counts removed as well as splitting the data into a data frame of counts and data frame of gene information. For this data set a total of 29458 genes were removed. 
```{r}
#Read in data set
sub.Data = read.table("assignment2data_subsampled_665968.tsv", header=TRUE, sep="\t")
dim(sub.Data)

#remove genes with 0 counts in male and female
sub.low.counts = apply(sub.Data[,1:20],1,median)==0 & apply(sub.Data[,21:40],1,median)==0
sub.filtered = sub.Data[!sub.low.counts,]

#split data
sub.expression = sub.filtered[,1:40]
sub.geneInfo = sub.filtered[,41:44]
```

The DGE object was created for edgeR and normalised. The MDS plot still shows separation between the male and female groups. 
```{r, eval=FALSE}
#create dge object, using same group vector from task 1
sub.dge = DGEList(sub.expression, group=ismale)
#normalise
sub.dge = calcNormFactors(sub.dge)
#plot MDS plot
plotMDS(sub.dge, col=c( rep("red",20), rep("blue",20) ))

```

The common and tag wise dispersion are estimated for the data set and the are viewed using the `plotBCV()` function. An exact test was used to find 
```{r, fig.cap='BCV plot for the sub samlped data.'}
#fit models
sub.dge = estimateCommonDisp(sub.dge)
sub.dge = estimateTagwiseDisp(sub.dge)
plotBCV(sub.dge)
```

To test for DGE  in the sum sampled data set the exact test was used to look for statistically significant differences between the means of the negative binomial distribution fitted to the data.
```{r, eval=FALSE}
sub.results = exactTest(sub.dge)
```

The `plotMeanVar()` function was used to visualise how the data is fitted to the models. 
```{r, fig.align='The mean variance plot showing the relationship between the mean and variance.'}
#plot mean variance plot to show how data fits to the model
plotMeanVar(sub.dge, show.raw.vars=TRUE, show.tagwise.vars=TRUE, 
            show.binned.common.disp.vars=FALSE, show.ave.raw.vars=FALSE, dispersion.method="qcml", 
            NBline=TRUE, nbins=100, pch=16 , 
            xlab="Mean Expression (Log10 Scale)", ylab = "Variance (Log10 Scale)")
```

Using the same methods as Task 1 we are able to find the to ten differential expressed genes as well as the bottom 10 genes. Using the `decideTestsDGE()` function again we can view the number of genes edgeR have significantly different fold changes. 
```{r}
#Sub top ten hits
sub.topGI = merge(sub.geneInfo[rownames(sub.topTen),], topTags(sub.results, n=10), by=0, sort=FALSE)
sub.topGI[,!(colnames(sub.topGI) %in% del), drop=FALSE]

#Sub bottom 10 hits
sub.low = topTags(sub.results, n=nrow(sub.results$table))$table
sub.low.GI = merge(sub.geneInfo[rownames(sub.low),], tail(sub.low, n=10), by=0, sort=FALSE)
sub.low.GI[,!(colnames(sub.low.GI) %in% del), drop=FALSE]

#Summary of up/down regulaton genes, tagwise results
sub.de = decideTestsDGE(sub.results)
summary(sub.de)
sub.detags = rownames(sub.dge)[as.logical(sub.de)]
```

The heat map shows still clear differences in the expressions between the groups but blocks are less clear and well defined. 
```{r, fig.cap='Heatmap for the subsampled data.'}
#heatmap
sub.topgenes = rownames(topTags(sub.results, n=20))
sub.log.dge = cpm(sub.dge, prior.count=2, log=TRUE)
sub.topgene.log.expression = sub.log.dge[sub.topgenes,]
heatmap(sub.topgene.log.expression, col = redtogreen)
```

Histogram showing the distribution of the P values for the sub sampled data set. 
```{r, fig.cap='Histogram of Pvalues of sub sampled data set.'}
#histogram
hist(sub.results$table$PValue, breaks=150)
```

Smear plot showing the fold changes for the genes in the sub sampled data set. 
```{r, fig.cap='Sub sampled smear plot'} 
sub.resultsTbl = topTags(sub.results, n=nrow(sub.results$table))$table
sub.de.genes = rownames(sub.resultsTbl)[sub.resultsTbl$PValue <= 0.05]
plotSmear(sub.results, de.tags=sub.de.genes)
abline(h=c(-2, 2), col="blue") 

```

## Task 2: Numerical Questions 
1. Number of genes with p-values below 0.05, What proportion of the genes tested?
```{r, strip.white=FALSE}
nrow(sub.results$table[sub.results$table$PValue < 0.05, ])
(nrow(sub.results$table[sub.results$table$PValue < 0.05, ]) / nrow(sub.results$table))
```

2. Out of all genes on Y chromosome, how many have a P-value below 0.05?
```{r, strip.white=FALSE}
sub.sig = sub.results$table[sub.results$table$PValue < 0.05, ]
sub.sigGI = sub.geneInfo[rownames(sub.sig),]
sub.sigY = sub.sigGI[sub.sigGI$chromosome_name=="Y", ]
nrow(sub.sigY)
```

3. Out of the 100 genes with the lowest P-value, how many are from the X chromosome?
```{r, strip.white=FALSE}
sub.top100 = topTags(sub.results, n=100)
sub.sigX = sub.geneInfo[rownames(sub.top100), ]
nrow(sub.sigX[sub.sigX$chromosome_name=="X", ])
```

4. What is the log-fold change for the gene XIST?
```{r, strip.white=FALSE}
subxist = sub.geneInfo[sub.geneInfo$gene_name=="XIST", ]
sub.xistResult = sub.results$table[rownames(sub.results$table)==rownames(xist), ]
round(sub.xistResult$logFC,digits=4)
```

### Task 3: Randomised sub sample set
The third task is to perform a sanity check and to be sure that our analysis does not contain any obvious errors or bias. As there is no reason to suspect significant differential expression between two randomly assigned groups we group individuals at random as opposed to their gender. 

To assign the groups randomly we use R to create a vector or random 0's and 1's. The set seed function ensures that we get the same randomisation each time the analysis is carried out. 
```{r, eval=FALSE}
#create vector of random 0 and 1
seed = 1234
set.seed(seed)
randomgroup = sample(c(0,1), 40, replace=TRUE)
```
```{r}
randomgroup
```

The same analysis is performed as task 2 on the data. The only difference is that we are now using the randomly assigned groups. 
```{r,eval=FALSE}
#create dge object
rand.dge = DGEList(sub.expression, group=randomgroup)
#normalise
rand.dge = calcNormFactors(rand.dge)
```

As there is still similarity between the male and females there are still clustered together in the MDS plot. 
```{r,fig.cap='MDS plot for the randomly assigned group.'}
#plot MDS plot
plotMDS(rand.dge, col=c( rep("red",20), rep("blue",20) ))
```

```{r,fig.cap='BCV plot for the randomly assigned groups.'}
#fit models
rand.dge = estimateCommonDisp(rand.dge)
rand.dge = estimateTagwiseDisp(rand.dge)
plotBCV(rand.dge)
rand.results = exactTest(rand.dge)
```

```{r,fig.cap='Mean variance plot for the randomly assigned groups.'}
#plot mean variance plot to show how data fits to the model
plotMeanVar(rand.dge, show.raw.vars=TRUE, show.tagwise.vars=TRUE, show.binned.common.disp.vars=FALSE, show.ave.raw.vars=FALSE, dispersion.method="qcml", NBline=TRUE, nbins=100, pch=16 , xlab="Mean Expression (Log10 Scale)", ylab = "Variance (Log10 Scale)")
```

As the groups were assigned at random there are still some differential expressed genes but not as many as found in Task 2. 
```{r}
#get top ten hits
rand.topGI = merge(sub.geneInfo[rownames(rand.topTen),], topTags(rand.results, n=10), by=0, sort=FALSE)
rand.topGI[,!(colnames(rand.topGI) %in% del), drop=FALSE]

#bottom 10 hits
rand.low = topTags(rand.results, n=nrow(rand.results$table))$table
rand.lowGI = merge(sub.geneInfo[rownames(rand.low),], tail(rand.low, n=10), by=0, sort=FALSE)
rand.lowGI[,!(colnames(rand.lowGI) %in% del), drop=FALSE]

#Summary of up/down regulaton genes, tagwise results
rand.de = decideTestsDGE(rand.results)
summary(rand.de)
rand.detags = rownames(rand.dge)[as.logical(rand.de)]
```

## Visualisation
As expected there is no real clustering between the differential expressed genes. The heat map looks to be uniform and random.
```{r, fig.cap='Heat map from the randomly assigned gropus'}
#heatmap
rand.topgenes = rownames(topTags(rand.results, n=20))
rand.log.dge = cpm(rand.dge, prior.count=2, log=TRUE)
rand.topgene.log.expression <- rand.log.dge[rand.topgenes,]
heatmap(rand.topgene.log.expression)
```

There are still peaks in the histogram by they are viewer. Over all the histogram is smaller and looks more uniform.
```{r, fig.cap='Histogram for the randomly assigned groups.'}
#histogram
hist(rand.results$table$PValue, breaks=150)
```

The smear plot also shows few genes that are differential expressed with large fold changes. The summary from before also only shows three differential expressed genes with significant fold changes. 
```{r, fig.cap='Smear plot for randomly assigned groups.'}
#smear plot - shows the relationship between concentraiton and the fold-change across the genes. 
rand.resultsTbl = topTags(rand.results, n=nrow(rand.results$table))$table
rand.de.genes = rownames(rand.resultsTbl)[rand.resultsTbl$PValue <= 0.05]
plotSmear(sub.results, de.tags=rand.de.genes)
abline(h=c(-2, 2), col="blue") 
```

## Task 3 Numerical Questons
1. Number of genes with p-values below 0.05, What proportion of the genes tested?
```{r, strip.white=FALSE}
nrow(rand.results$table[rand.results$table$PValue < 0.05, ])
(nrow(rand.results$table[rand.results$table$PValue < 0.05, ]) / nrow(rand.results$table))
```

2. Out of all genes on Y chromosome, how many have a P-value below 0.05?
```{r, strip.white=FALSE }
rand.sig = rand.results$table[rand.results$table$PValue < 0.05, ]
rand.sigGI = sub.geneInfo[rownames(rand.sig),]
rand.sigY = rand.sigGI[rand.sigGI$chromosome_name=="Y", ]
nrow(rand.sigY)
```

3. Out of the 100 genes with the lowest P-value, how many are from the X chromosome?
```{r, strip.white=FALSE}
rand.top100 = topTags(sub.results, n=100)
rand.sigX = sub.geneInfo[rownames(rand.top100), ]
nrow(rand.sigX[rand.sigX$chromosome_name=="X", ])
```

4. What is the log-fold change for the gene XIST?
```{r, strip.white=FALSE}
rand.xist = sub.geneInfo[sub.geneInfo$gene_name=="XIST", ]
rand.xist.Result = rand.results$table[rownames(rand.results$table)==rownames(xist), ]
rand.xist.Result$logFC
```

### Task 4: Discussion 
The purpose of this analysis was to find genes in our data set that were differential expressed between the two groups. As expected we found a number of differential expressed genes that are located on the Y chromosome. 

#Comparison of Task 1 and 2
The first task involved using high-depth RNA-seq data and the second involved the same groups but lower depth. The lower depth meant meant less statistical power for detecting DE genes. The first process of filtering counts we removed an extra 10,015 genes from the analysis. 

Comparing the two MDS for the higher and lower depth RNA-seq data we can see clearly in both the clustering between the male and female groups. So reduced depth does not effect how similar the similarity between the data. 

The dispersion estimates for the two data sets are very similar. Looking at the BCV plots for the two data sets they appear to be very similar the major difference being the number of tag wise estimates (as more genes were removed). Looking at the raw value for the common dispersion they are also very close. The common dispersion for high depth was `r dge$common.disperson` and for lower depth `r sub.dge$commondispersion`. 

As expected the genes in the top ten differential expressed list for both data sets are slightly different. As the counts are different slightly different results are given but common genes are seen in both lists. The same applies to the lowest DE genes. 

Looking at the decideTestsDGE results we can see there are fewer genes that are significantly DE. The higher coverage with more power has found more genes both negatively and positively DE. The heat maps show clear differential expression between the positively and negatively expressed genes in the higher depth compared to the lower depth. This is expected as we have removed genes reducing the power of our experiment. 

The comparison of histograms show the frequency of P value has reduced, this is again due to reduced number of genes in the lower coverage data set and the lower power of the analysis. Looking at the numerical analysis we are still detecting the same proportion of genes with P Value below 0.05. The log-fold change of the xist gene is also near the same value. 


###Comparison of Task 2 and Task 3
For Task 3 we performed some sanity checking by randomising the group assignment of the replicants. This in theory should show that there is no differential expressed genes between the two groups, showing us that our first analysis was working and detecting real DE genes. How ever this is random so we still do detect some differential expressed genes. 

The MDS plot still shows clear similarity between the male and female samples despite being assigned to groups randomly. The random assignments does not change the similarity between the samples. 

The comparison of heat maps clearly show the lack of differential expression between the randomly assigned groups. With no clear clustering of genes everything appears random. With the output of the decideTestsDGE from edgeR there are only three DE genes found compared to 46 from task 2.The smear plot also only shows the few genes that are above or below the 2x fold lines on the graph. 

As expected we have fewer genes that have a P Value lower then 0.05 as we are not detecting many differential expressed genes. Looking at the difference in log-fold changes in the xist gene between the two tasks. The log-fold change in Task 2 is -9.8, where in Task 3 it is 0.6323 showing the lack of detection of DE in randomly assigned groups. 


```{r}
#Session info
sessionInfo()
```